src/DispatchPlot.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2020, Battelle Energy Alliance, LLC
+# ALL RIGHTS RESERVED
+"""
+  Author: dylanjm
+  Date: 2021-05-18
+"""
+import os
+import sys
+import itertools as it
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+
+from PluginBaseClasses.OutStreamPlotPlugin import PlotPlugin, InputTypes, InputData
+
+# Matplotlib Global Settings
+plt.rc("figure", figsize=(12, 8), titleweight='bold')
+plt.rc(
+  "axes",
+  titleweight="bold",
+  labelsize=12,
+  axisbelow=True,
+  grid=True
+)
+plt.rc("savefig", bbox="tight")
+plt.rc("legend", fontsize=12)
+plt.rc(["xtick", "ytick"], labelsize=10)
+
+
+class DispatchPlot(PlotPlugin):
+
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Define the acceptable user inputs for this class.
+      @ In, None
+      @ Out, specs, InputData.ParameterInput,
+    """
+    specs = super().getInputSpecification()
+    specs.addSub(InputData.parameterInputFactory('source', contentType=InputTypes.StringType))
+    return specs
+
+  def __init__(self):
+    """
+      Constructor.
+      @ In, None
+      @ Out, None
+    """
+    super().__init__()
+    self.printTag = 'HERON.DispatchPlot'
+    self._sourceName = None
+    self._source = None
+
+  def handleInput(self, spec):
+    """
+      Reads in data from the input file
+      @ In, spec, InputData.ParameterInput, input information
+      @ Out, None
+    """
+    super().handleInput(spec)
+    for node in spec.subparts:
+      if node.getName() == 'source':
+        self._sourceName = node.value
+
+  def initialize(self, stepEntities):
+    """
+      Set up plotter for each run
+      @ In, stepEntities, dict, entities from the Step
+      @ Out, None
+    """
+    super().initialize(stepEntities)
+    src = self.findSource(self._sourceName, stepEntities)
+    if src is None:
+      self.raiseAnError(IOError, f'Source DataObject {self._sourceName} was not found in the Step!')
+    self._source = src
+
+  @staticmethod
+  def _group_by(iterable):
+    """
+      @ In, iterable, list, a list of column names to group-by.
+      @ Out, gr, dict, a dictionary containing a mapping of grouped variable names.
+    """
+    gr = {}
+    for var in iterable:
+      key = var.split('__')[-1]
+      if key in gr.keys():
+        gr[key].append(var)
+      else:
+        gr[key] = [var]
+    return gr
+
+  def run(self):
+    """
+      Generate the plot
+      @ In, None
+      @ Out, None
+    """
+    df = self._source.asDataset().to_dataframe().reset_index()
+    dispatch_vars = list(filter(lambda x: "Dispatch__" in x, df.columns))
+    grouped_vars = self._group_by(dispatch_vars)
+
+    # Loop through RAVEN_sample_ID variable
+    for sample_id in df.iloc[:, 0].unique():
+      # Loop through macro-step variable (i.e. YEAR)
+      for macro_step in df.iloc[:, 2].unique():
+        # Loop through _ROM_CLUSTER variable
+        for cluster in df.iloc[:, 3].unique():
```

Comment:
are the indices here always the same? That's quite interesting if so.

Reply:
I believe they will always be the same index but not the same name (ie 'YEAR' vs 'Year') hence why I'm using index here. 

The reason why I believe the indices will be constant is because they come from xarray as index variables in pandas. So they must be written to xarray in a specific order. If that makes sense?

Reply:
Nice, okay. That works for now, hard to foresee what the future may hold.